clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc

```diff
@@ -0,0 +1,66 @@
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+By default, a clustered Infinispan cache waits for the transfer of its initial state before it is available.
+The <state-transfer enabled="..."/> attribute was deprecated in WF10 with the expectation that it would be auto-enabled as appropriate based on the cache mode.
+However, several customers have complained that that their initial state transfer would timeout, and want the ability to disable it altogether.
+When the cache is sufficiently large, a synchronous state transfer isn't always feasible within the configured timeout, especially for a replicated-cache.
+
+Infinispan supports 3 options for configuring initial state transfer:
+. Enable initial state transfer.  In this mode, the state transfer is initiated and the cache only starts once the state transfer completes.  This ensures that a cache starts "warm".  If the state-transfer does not complete within a configured timeout, the cache startup aborts.
+. Disable initial state transfer.  In this mode, the cache populates with the appropriate data as it is accessed.  This is reasonable for a replicated cache, but can be dangerous for a distributed cache, since subsequent leavers can more easily result in data loss.
+. Async, non-blocking initial state transfer.  In this mode, the initial state transfer initiates but startup does not wait for its completion.  This is a compromise between the default behavior and the option above.   The cache will start cold, but the period of data loss vulnerability only lasts until the state transfer completes.  This is the best option for  large distributed caches.  For this option, a zero timeout would enable this behavior.
+
+This RFE proposed to resurrect the enabled attribute of state transfer and to additionally use asynchronous state transfer when the configured timeout is zero.
+
+== Issue Metadata
+
+=== Issue:
+
+* {issue-base-url}/EAP7-888
+
+=== Related Issues:
+
+* {issue-base-url}/WFLY-9648
+
+=== Dev Contacts:
+
+* mailto:{email}[{author}]
+
+=== QE Contacts:
+
+* mailto:mvinkler@redhat.com[Michal Vinkler]
+
+=== Affected Projects or Components:
+
+WildFly, Clustering
+
+=== Other Interested Projects:
+
+== Requirements
+
+=== Hard Requirements
+
+Un-deprecate the existing "enabled" attribute of the component=state-transfer resource, which was deprecated for WF10.
+Disable awaitInitialTransfer if state-transfer is enabled and timeout <= 0.
+
+When transforming the model for legacy slaves, for model versions between 4.0 and 6.0, exclusively, reject if the enabled attribute is false.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+== Test Plan
+
+Run standard set of failover smartfrog tests with state-transfer configured with a 0 timeout.
+Validate that async, non-blocking state transfer results does not introduce sampling errors.
+Generate visual feedback of performance impact of the additional remote gets following failover.
```

Comment:
So you would like to have a performance test with a failover?

Reply:
Surely having a performance test for the events where failover happens is needed. Ideally these should be the same as the original failover tests we have but obtain better representation of performance data.

Reply:
I guess we can run a standard failover test and collect statistics as we do in performance tests, output is a set of graphs (throughput, response time, etc.) you are familiar with.

Reply:
@mvinkler Correct.

Reply:
@pferraro 
`/subsystem=infinispan/cache-container=ejb/distributed-cache=dist/component=state-transfer:add(enabled=true,timeout=0)`
works correctly but outputs the following warning:
`12:23:05,030 INFO  [org.jboss.as.controller.management-deprecated] (pool-3-thread-1) WFLYCTL0028: Attribute 'enabled' in the resource at address '/subsystem=infinispan/cache-container=ejb/distributed-cache=dist/component=state-transfer' is deprecated, and may be removed in a future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.
{"outcome" => "success"}`
the same for:
`/subsystem=infinispan/cache-container=server/replicated-cache=default/component=state-transfer:add(enabled=true,timeout=0)
12:25:41,056 INFO  [org.jboss.as.controller.management-deprecated] (pool-3-thread-2) WFLYCTL0028: Attribute 'enabled' in the resource at address '/subsystem=infinispan/cache-container=server/replicated-cache=default/component=state-transfer' is deprecated, and may be removed in a future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.
{"outcome" => "success"}`

Reply:
@tommaso-borgato I'm no longer allowing state-transfer to be disabled - as this has unintended consequences.  This RFE has been limited to just enabling non-blocking initial state-transfer, via timeout=0.

Reply:
@pferraro: for the stress test with the state-transfer disabled, the configuration is `<state-transfer enabled="false"/>` correct?